### PR TITLE
[Evaluator Ergonomics] Shuffle Evaluator Infrastructure Out Of Requests 

### PR DESCRIPTION
**Pull request by CodaFi**
Shuffle extra infrastructure that was supporting requests out of the requests.  All requests should now consist of the requirements of a `SimpleRequest`.

In particular

- `WhereClauseOwner` owns the requirement iteration API now
- DefaultTypeRequest, while well-structured, contained a lot of unnecessary getters

Also drop the extra stdlib module ref kept around by the type checker for no reason if we can just grab this from the ASTContext.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeChecker.cpp [hunk 1]
```diff
@@ -209,17 +209,11 @@ Type TypeChecker::getObjectLiteralParameterType(ObjectLiteralExpr *expr,
...
-
-  if (!StdlibModule) {
-    return dc->getParentModule();
+  if (auto *stdlib = Context.getStdlibModule()) {
+    return stdlib;
   }
 
-  assert(StdlibModule && "no main module found");
-  return StdlibModule;
+  return dc->getParentModule();
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.cpp]: Is this reachable? If not, you can remove this method entirely and update all callers to just call Context.getStdlibModule()

Comment (CodaFi) [reply lib/Sema/TypeChecker.cpp]: It should come up if we're compiling stdlib itself.  But there are so few callers now it's something I'll look into.

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**